### PR TITLE
Switch to using hardhat-deploy for deployments

### LINKED ISSUE #43: Investigate and possibly switch to hardhat-deploy
https://github.com/wighawag/hardhat-deploy

There are two similar repos from the same person but this one seems to be the most up-to-date. They implement a ton of features we'd want. If it's stable and actually works then we should use it.

I'm dubious mainly because deploying using hardhat's signer is inconsistent. It tends to hang, hence EthersFastSubmitWallet.

### PR DESCRIPTION
**Pull request by TomAFrench**
I saw [this message](https://discordapp.com/channels/378030344374583298/378030691155312641/831347969235222528) in the discord and thought I might as well implement it.

`src/deployer.ts` has been nuked with some of it's config stuff moved into `tasks/deploy.ts` which is now just a wrapper around hardhat-deploy which then automatically verifies the contracts. 

I've tried not to mess too much with the surrounding code so I've left `addresses.ts` (and made sure that it's properly updated) although it would be probably best to make it read from the deployment json files created by hardhat-deploy as a source of truth in future.

One notable omission from this is that hardhat-deploy doesn't allow a customisable number of confirmations whereas `deployer.ts` did. I'm not sure what issues you've run into but I've done a couple of test deploys to kovan/mumbai and everything went well.

This will help address #158 as you can just fill in a skeleton deployment file for the external addresses for the networks you're deploying to and hardhat-deploy won't redeploy an identical contract.

closes #43

### CODE REVIEW

**Comment by TomAFrench** [comment]
Taking a second look at this I just noticed that all the `makeSigner` stuff is now unused (as I'm just pulling the address) so it can be removed and we should see if there's crucial in anything in there needs to be handled in the new deploy scripts.

Edit: It's also used in a number of other tasks so probably worth just keeping it for consistency.

File: packages/smart/test/bpool-test.ts [hunk 1]
```diff
@@ -1,4 +1,4 @@
-import { SignerWithAddress } from "@nomiclabs/hardhat-ethers/signers";
+import { SignerWithAddress } from "@nomiclabs/hardhat-ethers/dist/src/signer-with-address";
```
Comment (robert-e-davidson3) [inline packages/smart/test/bpool-test.ts:1]: Why this change? I'd like to avoid depending on `dist` directly.

**Comment by robert-e-davidson3** [comment]
Thanks for jumping on this!

The custom confirmations is for matic and arbitrum. Really it's a hack so I'm glad it can be cut.

I'm concerned that this won't work on arbitrum since it avoids using our custom wallet built in part to handle arbitrum's weirdness. But confirmations (or sleeps) were needed for mumbai before so maybe it isn't a problem.

This will need a refactor when #227 is merged in. (It's closed but its branch will be merged in when UI integration is done.)
Hopefully that gets in today but Monday is likely.

@justinbarry Thoughts on addresses.ts deprecation? I don't want to lose our typing but if there's a clean solution...

File: packages/smart/test/bpool-test.ts (hunk 1 above)
Comment (TomAFrench) [reply packages/smart/test/bpool-test.ts:1]: I've had to alias `@nomiclabs/hardhat-ethers` to `hardhat-deploy-ethers` for compatibility reasons (https://github.com/wighawag/hardhat-deploy#installation) and was getting ts errors unless I did this unfortunately.

**Comment by TomAFrench** [comment]
No worries :)

I'd say maybe not completely deprecate `addresses.ts` (simpler file to import into the UI rather than something which then goes and reads a bunch of json files) but I think ideally it should be generated as a whole from trawling the deployments folder rather than relying on updates at deploy-time. It's not a big deal though.

**Comment by justinbarry** [comment]
@TomAFrench This is amazing. Thank you!

Addresses.ts exists because we had some issues in the past with `tsc` watching json files and would end up in a weird state.  Precisely what you posited.

**Comment by TomAFrench** [comment]
This PR has been updated to account for #224.

**Comment by TomAFrench** [comment]
> I'm concerned that this won't work on arbitrum since it avoids using our custom wallet built in part to handle arbitrum's weirdness. But confirmations (or sleeps) were needed for mumbai before so maybe it isn't a problem.

@robert-e-davidson3 I've been having a bit of a look into arbitrum tonight to compare with Optimism so thought I'd check this out. I could deploy these to both public testnet and a local version so all seems good.

I've updated this PR to match the format that was introduced in #252 but the verify script for hardhat-deploy automatically pulls constructor args from the deployment files so you could probably remove those if you want.

**Comment by robert-e-davidson3** [comment]
I did a rebase to catch the interim changes in #387.
